src/librustc_middle/ty/relate.rs

```diff
@@ -510,7 +510,7 @@ pub fn super_relate_consts<R: TypeRelation<'tcx>>(
     let tcx = relation.tcx();
 
     let eagerly_eval = |x: &'tcx ty::Const<'tcx>| {
-        if !x.val.has_local_value() {
+        if !x.val.needs_infer() {
             return x.eval(tcx, relation.param_env()).val;
```

Comment:
cc @oli-obk @varkor This seems unnecessary, `ty::Const::eval` handles `needs_infer` itself.